### ISSUE TITLE
Mac ,FPS drop from 60 to 30 on FullScreen .

### ISSUE THREAD

**Issue by Kotaiba** [cocos2d/cocos2d-x#16561]
- cocos2d-x version: 3.11.1
- devices test on: Macbook Retina 15 , MAC OSX
- developing environments
  - Xcode version: Version 7.3.1 

This Issue only on  `glview = GLViewImpl::createWithFullScreen("DCR");`
if i use this the game will run 60 fps `glview = GLViewImpl::createWithRect("DCR", Rect(0, 0, 1136 ,640));`
On some mac devices the fps will drop from 60 to 30 . and go back to 60 again 
the game drop even on the  loading that have only 1 sprite the logo.
please see the attached video . it will show the issue ,
[video-1473507408.mp4.zip](https://github.com/cocos2d/cocos2d-x/files/465879/video-1473507408.mp4.zip)

**Comment by slackmoehrle** [comment]
I watched the video. Interesting. But can you clarify a bit? Which line are you using that cases the drop? Also, by Mac devices do you you mean iphones and ipads or desktop OSX?

**Comment by Kotaiba** [comment]
Only on desktop OSX , the drop happen only when the game is on FullScreen 
`glview = GLViewImpl::createWithFullScreen("DCR");`

**Comment by dumganhar** [comment]
If the issue only appears on macOS desktop, it's probably an issue of GLFW in fullscreen.

**Comment by Kotaiba** [comment]
@dumganhar  let me do more test on windows .

**Comment by dumganhar** [comment]
@Kotaiba, that will be better for us to check the issue. Thanks. :)

**Comment by Kotaiba** [comment]
@dumganhar Same issue on windows but it drop to 20-30-60fps only on FullScreen.

**Comment by dumganhar** [comment]
@Kotaiba, interesting. I will try to reproduce it.

**Comment by dumganhar** [comment]
@Kotaiba , I tried cpp-empty-test on my MacBook Pro (Retina, 15-inch, Early 2013). And I could not reproduce it. Could you test cpp-empty-test?

**Comment by Kotaiba** [comment]
@dumganhar  cpp-empty-test  worked fine . but try to run cpp-tests on fullscreen and open basic-shaders it sometimes happen see the attached image . i am using MacBook Pro (Retina, 15-inch, Mid 2014)
<img width="1440" alt="screen shot 2016-09-13 at 11 43 20 am 1" src="https://cloud.githubusercontent.com/assets/3635272/18467126/5e0b1a5a-79a7-11e6-9725-ac56c8b9356d.png">

question: why when i set the fps to 120 it stay on 60 ?

**Comment by dumganhar** [comment]
Yep, I reproduced it by Shader test.
And even though I set fps to 120, it still kept 30 fps on my Mac.
Probably, while in full screen of `Frag shader test`, GPU is too busy. And V-Sync takes effect and makes CPU to wait GPU to finish its operation. My description may be wrong.
I asked @dabingnn who is an expert of Rendering for help. And he said fps jump from 60 to 30 in fullscreen is reasonable while GPU is too busy.

@ricardoquesada , do you have any idea about this?

**Comment by Kotaiba** [comment]
@dumganhar i mean when i set the fps to 120 it stay at 60 not on this issue ,in general .but on some windows devices it go up to 120 and other it stay on 60 .?

**Comment by dumganhar** [comment]
@Kotaiba , sorry, I have no idea about that now.
Maybe @ricardoquesada could help. :)

**Comment by ricardoquesada** [comment]
For me, cpp-tests was working at 60FPS in most of the tests. Some run at 15, specially the shader ones.
I guess it is because in full screen it has "more pixels" to fill ? But it created a 1440x900 screen which is not that big, specially compared to iPhones and other 6" devices.
My guess is that it could be a bug either in GLFW or in the Mac OpenGL driver... or perhaps we are setting it incorrectly? I've just debugged it and I couldn't find anything strange in our code.

**Comment by Bilalmirza** [comment]
Hey,
@ricardoquesada @dumganhar  I had the same issue, when running fullscreen on xcode. Fps would bounce between 30-60.

Instead go to debug folder and run .app file. It gives much better fps for me its 58-60.

Thanks

**Comment by Kotaiba** [comment]
Debug folder did't solve it for me . and i tied to export it from Archive same issue .
It's weird how good the game running as popup ,and how bad its on fullscreen the hole game is lagging not only on the shaders .even on the loading. 

if you run cpp-tests on fullscreen you can see the fps is not as stable as the popup not only on the shader test .

**Comment by Bilalmirza** [comment]
@Kotaiba  You are right... i tested it on a 2 year old macbook and it was lagging pretty bad.

**Comment by Bilalmirza** [comment]
@minggo  Hi, Previously my game was supposed to launch on android 6months ago. But our team realised that android had too many performance and audio issues.

So we planned to go desktop first.

Mac performance is much better than android but it is not as good as iphone version.

Here is a Vid of A empty scene in full screen mac!

https://www.dropbox.com/s/dfl5bzg5baplxh0/macLag.mov?dl=0

the fps spikes alot in an empty scene.

My game runs 60 fps with 200 drawcalls and 600+ nodes spawned at all time on IPHONE.
Mac should be able to generate 180+ fps with no problem.

Mac Spec for the test.
Macbook pro 15-inch mid 2015
2.5ghz i7
Graphic card AMD r9 370x

OSX version 10.11.6

What can I do to get a better performance from mac Full Screen version?

Any help would be greatly appreciated thanks :D

**Comment by Rubonnek** [comment]
I was reading the developer log of Banished and the developer mentioned a similar FPS drop issue in the [entry of June 23, 2016](http://www.shiningrocksoftware.com/2016-06-23-passing-values/) while working on the Mac OS X port of the game. The last line of the log summarizes the root cause of the problem:

> ... don’t pass uniform arrays and large structs to a function by value in GLSL.

I think this might be the cause for the FPS drop in Cocos2D-X too -- a quick search of the `inout` keyword under the cocos directory returns empty:

```
$ cd ${COCOS_GIT_ROOT}/cocos
$ find . -type f -exec fgrep 'inout' {} + 
```

**Comment by minggo** [comment]
Why did you find `inout`?

**Comment by Rubonnek** [comment]
That's the change the Banished developer made to the `SkinPosition` function to fix the FPS drop in the Mac OS X port. That same keyword, which strictly speaking is a parameter qualifier, also appears in the [OpenGL ES Shading Language](https://www.khronos.org/files/opengles_shading_language.pdf). And there are [online references](http://www.shaderific.com/glsl-qualifiers/) that state that using this `inout` qualifier will make the parameter to be passed by reference in GLSL. In other words, the lack of that `inout` qualifier means that we are not passing anything by reference in GLSL and this could account for the FPS drop that we are experiencing. That's why I looked looked for it in the source code.

We are also lacking that `inout` qualifier in the Mandelbrot shader -- in fact, we don't have it in any of the shaders in cpp-tests:

```
$ cd ${COCOS_GIT_ROOT}/tests/cpp-tests/Resources
$ find . -type f -exec fgrep 'inout' {} +    # returns nothing
```

Perhaps by modifying the Mandelbrot shader to include that `inout` qualifier we might be able to test if this is issue can be fixed by including that qualifier somewhere else in the shaders where possible.

I might be wrong, I'm not an expert on GLSL and I don't know if this can be done, but I thought this might be worth looking into.

Update: Clarification.

**Comment by Bilalmirza** [comment]
@minggo hey, will this be looked into for 3.14 release or later?

**Comment by minggo** [comment]
It will be in later version.

**Comment by Bilalmirza** [comment]
Okay thanks :)

**Comment by RustyMoyher** [comment]
I've been upgrading to 3.13.1 this week and I am currently wrestling with the same issue. However, I'm seeing the 60 --> 30 --> 60 bounce in windowed mode too.

**Comment by Kotaiba** [comment]
@RustyMoyher is it only on Mac or both Mac and PC ?.

**Comment by RustyMoyher** [comment]
@Kotaiba Nope, only on Mac. Just finished testing in various Win 10 and Win 7 and in windowed and fullscreen modes: ran 60 fps without issue.

**Comment by Kotaiba** [comment]
@minggo  any update about the fix ?

**Comment by minggo** [comment]
Sorry, not yet. Because Mac is not a release platform, so i put it in low priority.

**Comment by minggo** [comment]
@Kotaiba I tested cpp-tests, the `Mandelbrot` shader test will drop to 30 fps. And other test cases don't have the issue. Did you mean the Mandelbrot shader test is an issue?

**Comment by Kotaiba** [comment]
@minggo  No the issue happen on my game 90% of the time when the game is on full screen, not only on the shaders .see the attached video it will show the issue clearly .
[IMG_2350.zip](https://github.com/cocos2d/cocos2d-x/files/860776/IMG_2350.zip)

**Comment by minggo** [comment]
Can not download it currently because of network speed. How can i reproduce it?

**Comment by Kotaiba** [comment]
the Mandelbrot shader test ,is one way to reproduce it .

**Comment by Kotaiba** [comment]
try to set on-off button to switch between fullscreen inside the shader test ,you will see how the fps limit it self to 30 ,and jump back to 60 on popup mode.

**Comment by minggo** [comment]
Yep, Mandelbrot shader test can reproduce it. But because it uses a specific shader and other shaders don't have the issue. So it is not a good test case. I wonder if it is shader related or not.

**Comment by Kotaiba** [comment]
I have 4 scenes on my game and the issue happen in all of them ,the scene on the video that i attached got spine animation and only few sprites.

**Comment by RustyMoyher** [comment]
@kaishiqi Have you seen this? https://github.com/glfw/glfw/issues/772

The fun technical dive on this issue can be found here: http://stackoverflow.com/questions/27758027/sprite-kit-serious-fps-issue-in-full-screen-mode-on-os-x

**Comment by minggo** [comment]
@RustyMoyher thanks for the link. Interesting issue. I think we should wait glfw to fix the issue, or you guys can modify the glfw source code to have the patch which i don't think it is a elegant way to fix it.

**Comment by RustyMoyher** [comment]
I don't think the bug I linked to was with glfw. It a bug with macOS.

**Comment by minggo** [comment]
Yep, you are right. I mean if we want to work around, we should modify glfw source codes, but it is not a good way.

**Comment by Bilalmirza** [comment]
@minggo will this get fixed?

**Comment by Bilalmirza** [comment]
@Kotaiba were you able to fix this?

**Comment by Kotaiba** [comment]
@Bilalmirza  no , this is not cocos2d bug . 
For me i made 3 sizes for mac and the user can change between them and full screen is not one of them .  view->setWindowed(1136 ,640);

**Comment by Bilalmirza** [comment]
@Kotaiba thanks!
